book/src/dev/state-db-upgrades.md

```diff
@@ -1,18 +1,63 @@
 # Zebra Cached State Database Implementation
 
+## Current Implementation
+
+### Verification Modes
+[verification]: #verification
+
+Zebra's state has two verification modes:
+- block hash checkpoints, and
+- full verification.
+
+This means that verification uses two different codepaths, and they must produce the same results.
+
+By default, Zebra uses as many checkpoints as it can, because they are more secure against rollbacks
+(and some other kinds of chain attacks). Then it uses full verification for the last few thousand
+blocks.
+
+When Zebra gets more checkpoints in each new release, it checks the previously verified cached
+state against those checkpoints. This checks that the two codepaths produce the same results.
+
 ## Upgrading the State Database
+[upgrades]: #upgrades
 
 For most state upgrades, we want to modify the database format of the existing database. If we
 change the major database version, every user needs to re-download and re-verify all the blocks,
 which can take days.
 
+### Writing Blocks to the State
+[write-block]: #write-block
+
+Zebra's state has two methods of getting to the latest format:
+- syncing from empty with the latest format, and
+- upgrading from a previous format to the latest format.
+
+This means that writing blocks uses two different codepaths, and they must produce the same results.
+This code is high risk, because discovering bugs is tricky, and fixing bugs can require a full reset
+and re-write of an entire column family.
+
+Most Zebra instances will do an upgrade, because they already have a cached state, and upgrades are
+faster. But we run a full sync in CI every week, because new users use that codepath. (And it is
+their first experience of Zebra.)
+
+When Zebra starts up and shuts down (and periodically in CI tests), we run checks on the state
+format. This makes sure that the two codepaths produce the same state on disk.
+
+To reduce code and testing complexity:
+- when a previous Zebra version opens a newer state, the entire state is considered to have that lower version, and
+- when a newer Zebra version opens an older state, each required upgrade is run on the entire state.
+
 ### In-Place Upgrade Goals
+[upgrade-goals]: #upgrade-goals
 
+Here are the goals of in-place upgrades:
 - avoid a full download and rebuild of the state
 - the previous state format must be able to be loaded by the new state
```

Comment:
Nitpick/Optional:

```suggestion
- Zebra's state must be able to upgrade data from the previous major version of its disk format
  and load data from the previous minor or patch versions of its disk format
```

Reply:
Thank you for the detailed review and feedback!

This really good feedback, even though it is not quite right - that means we need to clearly document the current behaviour.

```suggestion
- Zebra must be able to upgrade the format from previous minor or patch versions of its disk format
  (Major disk format versions are breaking changes. They create a new empty state and re-sync the whole chain.)
```